cmd/soroban-rpc/internal/methods/get_events.go

```diff
@@ -0,0 +1,475 @@
+	}
+	if event.ContractId == nil {
+		return false
+	}
+	needle := hex.EncodeToString((*event.ContractId)[:])
+	for _, id := range e.ContractIDs {
+		if id == needle {
+			return true
+		}
+	}
+	return false
+}
+
+func (e *EventFilter) matchesTopics(event xdr.ContractEvent) bool {
+	if len(e.Topics) == 0 {
+		return true
+	}
+	v0 := event.Body.MustV0()
+	for _, topicFilter := range e.Topics {
+		if topicFilter.Matches(v0.Topics) {
+			return true
+		}
+	}
+	return false
+}
+
+type TopicFilter []SegmentFilter
+
+func (t *TopicFilter) Valid() error {
+	if len(*t) < 1 {
+		return errors.New("topic must have at least one segment")
+	}
+	if len(*t) > 4 {
+		return errors.New("topic cannot have more than 4 segments")
+	}
+	return nil
+}
+
+func (t TopicFilter) Matches(event []xdr.ScVal) bool {
+	for _, segmentFilter := range t {
+		if segmentFilter.wildcard != nil {
+			switch *segmentFilter.wildcard {
+			case "*":
+				// one-segment wildcard
+				if len(event) == 0 {
+					// Nothing to match, need one segment.
+					return false
+				}
+				// Ignore this token
+				event = event[1:]
+			default:
+				panic("invalid segmentFilter")
+			}
+		} else if segmentFilter.scval != nil {
+			// Exact match the scval
+			if len(event) == 0 || !segmentFilter.scval.Equals(event[0]) {
+				return false
+			}
+			event = event[1:]
+		} else {
+			panic("invalid segmentFilter")
+		}
+	}
+	// Check we had no leftovers
+	return len(event) == 0
+}
+
+type SegmentFilter struct {
+	wildcard *string
+	scval    *xdr.ScVal
+}
+
+func (s *SegmentFilter) UnmarshalJSON(p []byte) error {
+	s.wildcard = nil
+	s.scval = nil
+
+	var tmp string
+	if err := json.Unmarshal(p, &tmp); err != nil {
+		return err
+	}
+	if tmp == "*" {
+		s.wildcard = &tmp
+	} else {
+		var out xdr.ScVal
+		if err := xdr.SafeUnmarshalBase64(tmp, &out); err != nil {
+			return err
+		}
+		s.scval = &out
+	}
+	return nil
+}
+
+type PaginationOptions struct {
+	Cursor string `json:"cursor,omitempty"`
+	Limit  uint   `json:"limit,omitempty"`
+}
+
+type EventStore struct {
+	Client TransactionClient
+}
+
+type TransactionClient interface {
+	Transactions(request horizonclient.TransactionRequest) (horizon.TransactionsPage, error)
+}
+
+// TODO: Extract this to a new package 'eventid'
+// Build a lexically order-able id for this event record. This is
+// based on Horizon's db2/history.Effect.ID method.
+type EventID struct {
+	*toid.ID
+	EventOrder int32
+}
+
+// String returns a string representation of this id
+func (id EventID) String() string {
+	return fmt.Sprintf(
+		"%019d-%010d",
+		id.ToInt64(),
+		id.EventOrder+1,
+	)
+}
+
+func (id *EventID) Parse(input string) error {
+	parts := strings.SplitN(input, "-", 2)
+	if len(parts) != 2 {
+		return fmt.Errorf("invalid event id %s", input)
+	}
+
+	// Parse the first part (toid)
+	idInt, err := strconv.ParseInt(parts[0], 10, 64)
+	if err != nil {
+		return errors.Wrapf(err, "invalid event id %s", input)
+	}
+	parsed := toid.Parse(idInt)
+	id.ID = &parsed
```

Comment:
super minor, is it worthwhile to only update the attributes of `id` instance at the end when err is nil, otherwise could return err here but have partially populated `id` instance with some attributes set.